File: src/main/java/io/gitlab/intended/storagenetworks/client/registry/ScreenRegistry.java
package io.gitlab.intended.storagenetworks.client.registry;

import io.gitlab.intended.storagenetworks.client.gui.screen.CraftingTerminalScreen;
import io.gitlab.intended.storagenetworks.registry.BlockRegistry;
import net.fabricmc.fabric.api.client.screen.ContainerScreenFactory;
import net.fabricmc.fabric.api.client.screen.ScreenProviderRegistry;
import net.minecraft.container.Container;
import net.minecraft.util.Identifier;

public final class ScreenRegistry {

    public static void register() {
        register(BlockRegistry.CRAFTING_TERMINAL.ID, CraftingTerminalScreen::new);
    }

    protected static <C extends Container> void register(Identifier id, ContainerScreenFactory<C> factory) {
        ScreenProviderRegistry.INSTANCE.registerFactory(id, factory);
    }

}

File: src/main/kotlin/io/gitlab/intended/storagenetworks/gui/screen/ScreenRegistry.kt
package io.gitlab.intended.storagenetworks.gui.screen

import io.gitlab.intended.storagenetworks.block.BlockRegistry
import io.gitlab.intended.storagenetworks.block.ModBlock
import io.gitlab.intended.storagenetworks.gui.container.MasterContainer
import io.gitlab.intended.storagenetworks.gui.container.RequestContainer
import net.fabricmc.fabric.api.client.screen.ContainerScreenFactory
import net.fabricmc.fabric.api.client.screen.ScreenProviderRegistry
import net.minecraft.client.gui.screen.ingame.ContainerScreen
import net.minecraft.container.Container

object ScreenRegistry {

    fun init() {
        r(BlockRegistry.REQUEST) { i: RequestContainer -> RequestScreen(i) }
        r(BlockRegistry.MASTER) { i: MasterContainer -> MasterScreen(i) }
    }

    private fun <C : Container, S : ContainerScreen<C>> r(modBlock: ModBlock, function: (C) -> S) {
        ScreenProviderRegistry.INSTANCE.registerFactory(modBlock.id, ContainerScreenFactory(function))
    }

}
